edk2toollib/windows/policy/firmware_policy.py

```diff
@@ -0,0 +1,589 @@
+
+        errorMessage = 'Reserved2 not fully supported'
+        if (STRICT is True):
+            raise Exception(errorMessage)
+
+    def PopulateFromFileStream(self, fs, vtoffset=0):
+        if fs is None:
+            raise Exception('Invalid File stream')
+        (self.ObjectType, self.Element, self.OffsetToValue) = struct.unpack(
+            self.StructFormat, fs.read(self.StructSize))
+
+        errorMessage = 'Reserved2 PopulateFromFileStream does not deserialize Reserved2 values'
+        print(errorMessage)
+        if (STRICT is True):
+            raise Exception(errorMessage)
+
+    def Print(self, prefix=''):
+        print('%sReserved2' % prefix)
+        print('%s  ObjectType:  %x' % (prefix, self.ObjectType))
+        print('%s  Element:  %x' % (prefix, self.Element))
+        print('%s  ValueOffset:  %x' % (prefix, self.OffsetToValue))
+
+    def Serialize(self, ruleOut, valueOut=None, offsetInVT=0):
+        ruleOut += struct.pack(self.StructFormat, self.ObjectType, self.Element, self.OffsetToValue)
+        errorMessage = 'Reserved2 value serialization not supported'
+        print(errorMessage)
+        if (STRICT is True):
+            raise Exception(errorMessage)
+
+
+class FirmwarePolicy(object):
+    """
+    Class for storing, serializing, deserializing, & printing Firmware Policy structures
+    Typically handles privitive types iteself, or delegates to other classes
+    for non-primitive structures, e.g. Rule, PolicyValue, PolicyString
+    """
+    FixedStructFormat = '<HI16sHIHH'  # omits completely unsupported Reserved1 array
+    FixedStructSize = struct.calcsize(FixedStructFormat)
+
+    POLICY_BLOB_MIN_SIZE = 32  # bytes
+    POLICY_FORMAT_VERSION = 2
+    POLICY_VERSION = 1
+    POLICY_PUBLISHER = uuid.UUID('5AE6F808-8384-4EB9-A23A-0CCC1093E3DD')  # Do NOT change
+    FW_POLICY_ROOT_KEY = 0xEF100000
+    FW_POLICY_SUB_KEY_NAME_TARGET = 'Target'
+    FW_POLICY_SUB_KEY_NAME = 'UEFI'
+    FW_POLICY_VALUE_NAME = 'Policy'
+    FW_POLICY_TYPE = PolicyValueType.POLICY_VALUE_TYPE_QWORD
+    FW_POLICY_VALUE_DEFINED_MASK = 0x00000000FFFFFFFF
+    FW_POLICY_VALUE_OEM_MASK = 0xFFFFFFFF00000000
+    FW_POLICY_VALUE_ACTIONS_MASK = 0x0000FFFF0000FFFF
+    FW_POLICY_VALUE_STATES_MASK = 0xFFFF0000FFFF0000
+
+    """Defined Policy Actions"""
+    FW_POLICY_VALUE_ACTION_SECUREBOOT_CLEAR = 0x0000000000000001
+    FW_POLICY_VALUE_ACTION_TPM_CLEAR = 0x0000000000000002
+
+    """Defined Policy States"""
+    FW_POLICY_VALUE_STATE_DISABLE_SPI_LOCK = 0x0000000000010000
+
+    def __init__(self, fs=None):
+        if fs is None:
+            self.FormatVersion = self.POLICY_FORMAT_VERSION
+            self.PolicyVersion = self.POLICY_VERSION
+            self.PolicyPublisher = self.POLICY_PUBLISHER
+            self.Reserved1Count = 0
+            self.Reserved1 = []
+            self.OptionFlags = 0
+            self.Reserved2Count = 0
+            self.RulesCount = 0
+            self.Reserved2 = []
```

Comment:
This is confusing. This is a list of Reserved2 objects, but Reserved1 is a list of... something else?

Reply:
There are 2 arrays of different types that are not used in the Firmware Policy, Reserved1[] and Reserved2[].  When the Reserved1Count and Reserved2Count are both equal to 0, then the arrays are not present in the structure.  Cannot eliminate the Counts as they will be expected in the structure server-side.

Reply:
I have 2 unit tests for Reserved2.  The first one uses an old, non firmware policy that included Reserved2 entries.  The second one is synthetic, exercising some unused methods to ensure that they are basically functioning for future use.  I'm closing this for now, but re-open if you have strong concern here.  Still working on some other issues.